gslib/commands/iam.py

```diff
@@ -363,8 +374,158 @@ def get_gcloud_storage_args(self):
+
+    # N.B.: self.sub_opts stops taking in options at the first non-flagged
+    # token. The rest of the tokens are sent to self.args. Thus, in order to
+    # handle input of the form "-d <binding> <binding> <url>", we will have to
+    # parse self.args for a mix of both bindings and CloudUrls. We are not
+    # expecting to come across the -r, -f flags here.
+    it = iter(self.args)
+    for token in it:
+      if (STORAGE_URI_REGEX.match(token) and
+          IsKnownUrlScheme(GetSchemeFromUrlString(token))):
+        url_pattern_strings.append(token)
+        break
+      if token == '-d':
+        try:
+          patch_bindings_tuples.append(BindingStringToTuple(False, next(it)))
+        except StopIteration:
+          raise CommandException(
+              'A -d flag is missing an argument specifying bindings to remove.')
+      else:
+        patch_bindings_tuples.append(BindingStringToTuple(True, token))
+    if not patch_bindings_tuples:
+      raise CommandException('Must specify at least one binding.')
+
+    # All following arguments are urls.
+    for token in it:
+      url_pattern_strings.append(token)
+
+    url_objects = list(map(StorageUrlFromString, url_pattern_strings))
+    _RaiseErrorIfUrlsAreMixOfBucketsAndObjects(url_objects,
+                                               self.recursion_requested)
+
+    return patch_bindings_tuples, url_objects
+
+  def _run_ch_subprocess(self, args, stdin=None):
+    _, command = self._get_full_gcloud_storage_execution_information(args)
+    process = subprocess.run(
+        command,
+        env=self._get_shim_command_environment_variables(),
+        input=stdin,
+        stderr=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+        text=True,
+    )
+
+    if process.returncode != 0:
+      if self.continue_on_error:
+        self.logger.debug(process.stderr)
+      else:
+        raise CommandException(process.stderr)
+    return process
+
+  def run_gcloud_storage(self):
+    if self.sub_command != 'ch':
+      return super().run_gcloud_storage()
+
+    self.ParseSubOpts()
+    bindings_tuples, patterns = self._GetSettingsAndDiffs()
+    resource_type = self._get_resource_type(patterns)
+
+    list_settings = []
+    if self.recursion_requested:
+      list_settings.append('-r')
+
+    return_code = 0
+    for url_pattern in patterns:
+      self._RaiseIfInvalidUrl(url_pattern)
+
+      if resource_type == 'objects':
+        ls_process = self._run_ch_subprocess(
+            ['alpha', 'storage', 'ls', '--json'] + list_settings +
+            [str(url_pattern)])
+        if ls_process.returncode:
+          return_code = 1
+          continue
+
+        ls_output = json.loads(ls_process.stdout)
+        urls = [
+            resource['url']
+            for resource in ls_output
+            if resource['type'] == 'cloud_object'
+        ]
+      else:
+        urls = [str(url_pattern)]
+
+      # With the -r flag, container headings are printed ending in '/:'.
```

Comment:
Is this comment still necessary with JSON formatting?

Reply:
Good catch!